File: 2023/SystemDesign/MovingTotal.java
import java.util.HashMap;
import java.util.ArrayList;

public class MovingTotal {
    public static ArrayList<Integer> queue;
    public static int front; 
    public static HashMap<Integer, Integer> hashMap; 
    public static int initialSum;

    public MovingTotal() {
        queue = new ArrayList<Integer>();
        front = 0;
        hashMap = new HashMap<Integer, Integer>();
        initialSum = 0;
    
    }

    public ArrayList<Integer> getQueue() {
        return queue;
    }

    public HashMap getHashMap() {
        return hashMap;
    }

    public void add(int[] nums) {
        for (int i: nums) {
            queue.add(i);
            if ((queue.size() - front) >= 3 && hashMap.isEmpty()) {
                for (int j: queue) {
                    initialSum += j;
                }
                hashMap.put(initialSum, null);
                front++;
            } else if ((queue.size() - front) >= 3) {
                System.out.println(i);
                initialSum = initialSum - queue.get(front-1) + queue.get(queue.size()-1);
                hashMap.put(initialSum, null);
                front++;
            }
        }
    }

    public Boolean contains(int value) {
        return hashMap.containsKey(value);
    }
}

File: 2023/SystemDesign/MovingTotal.py
class MovingTotal:
    def __init__(self):
        self.queue = []
        self.dict = {}
        self.front = 0
        self.initial_sum = 0

    def add(self, nums):
        for i in nums:
            self.queue.append(i)
            if len(self.queue) >= 3 and not self.dict:
                self.initial_sum = sum(self.queue)
                self.dict[self.initial_sum] = 0
                self.front += 1
            elif len(self.queue) >= 3:
                self.initial_sum = self.initial_sum - self.queue[self.front-1] + self.queue[-1]
                self.dict[self.initial_sum] = 0
                self.front += 1
    
    def contains(self, value):
        return value in self.dict
    
    def __str__(self):
        return f"Queue: {self.queue}\nHashMap: {self.dict}"

moving_total = MovingTotal()
moving_total.add([1,2,3,4,5,6,7,8])
print(moving_total)
            